modules/monitoring-exposing-custom-application-metrics-for-horizontal-pod-autoscaling.adoc

```diff
@@ -0,0 +1,223 @@
+// Module included in the following assemblies:
+//
+// monitoring/monitoring.adoc
+
+[id='exposing-custom-application-metrics-for-horizontal-pod-autoscaling-{context}']
+= Exposing custom application metrics for horizontal pod autoscaling
```

Comment:
As written, I don't think that you have enough information to publish this module. Please get more information to fill in your questions.

Reply:
Agree. This seems rather not follow a specific example like "I want to autoscale my application with the number of HTTP request that come in". @s-urbaniak would you mind helping @rh-max with that, please?

Reply:
The specifics are not in the name of the procedure, but the example _is_ specific, and it is about the metric of number of HTTP requests. So, @s-urbaniak already gave all necessary information. Or do you think that some information is missing, @sichvoge ?

Reply:
I gave @rh-max a specific example for the custom metric `http_requests_total`. To generalize this one could describe a more generalized procedure to create the prometheus adapter configmap with the following steps:

1. Identify the custom metrics to be exposed (`http_requests_total` in the specific example).
2. Formulate a series query to let the prometheus adapter discover that metric in Prometheus, example: https://github.com/openshift/openshift-docs/pull/14149/files#diff-8a75b2e3ebc241ba05379f54d182cca7R117

```
seriesQuery: 'http_requests_total{namespace!="",pod!=""}'
```

3. Map the series labels to kubernetes resources: https://github.com/openshift/openshift-docs/pull/14149/files#diff-8a75b2e3ebc241ba05379f54d182cca7R118

```
        resources:
          overrides:
            namespace: {resource: "namespace"}
            pod: {resource: "pod"}
            service: {resource: "service"}
```

4. (optional): apply renaming configuration if semantics of the metric changes: https://github.com/openshift/openshift-docs/pull/14149/files#diff-8a75b2e3ebc241ba05379f54d182cca7R123

```
        name:
          matches: "^(.*)_total"
          as: "${1}_per_second"
```

5. Formulate the actual query to retrieve actual metrics data: https://github.com/openshift/openshift-docs/pull/14149/files#diff-8a75b2e3ebc241ba05379f54d182cca7R126

```
        metricsQuery: 'sum(rate(<<.Series>>{<<.LabelMatchers>>}[2m])) by (<<.GroupBy>>)'
```

*Note*: All of the above is subject to dramatic change in the course of refactorings/evolution of the prometheus adapter in future versions of OpenShift.